src/planner.cc

```diff
@@ -663,175 +668,205 @@ void Planner::Impl::bring_path_to_halt() {
-      segment->planned.v2 = segment_end_speed;
-    } else {
-      segment->planned.decel = 0;
-      segment->planned.travel = delta_steps;
-      segment->planned.v1 = segment->planned.v2 = segment->target.speed;
-    }
-
-    // Update variables for next loop cycle.
-    defining_axis = segment->target.defining_axis;
-    // In the next iteration segment, our speed should be the final speed
-    // which is the smallest between the highest speed we could reach or
-    // the current segment start_speed (the starting segment is special).
-    speed = buffer_index && (buffer_index != last_planned_invariant_segment_)
-              ? std::min(segment->planned.v1, segment->target.start_speed)
-              : segment->planned.v0;
-
-    // Let's reset the final speed always as v1. We want to keep a consistent
-    // profile state which means if no accel -> v0 == v1 and no decel -> v1 ==
-    // v2.
-    segment->planned.v0 = segment->planned.v1;
-    segment->planned.accel = 0;
+    next_segment = previous_segment;
   }
+  bool first_accel = false;
+  previous_segment = nullptr;
 
-  // Compute the forward pass.
+  // Perform a forward pass.
   for (++buffer_index; buffer_index < (int)planning_buffer_.size();
        ++buffer_index) {
-    segment = planning_buffer_[buffer_index];
-    const auto speed_factor =
-      get_speed_factor_for_axis(&segment->target, defining_axis);
-    const double segment_start_speed =
-      (speed_factor != 0) ? speed / speed_factor : 0;
-    defining_axis = segment->target.defining_axis;
-    segment->planned.v0 = segment_start_speed;
-
-    // Recompute the accel ramp if necessary.
-    if (segment->planned.v1 == 0) {
-      speed = 0;
-      continue;
-    }
-
-    // Nothing to accelerate. Starting speed is above the travel speed.
-    if (segment_start_speed >= segment->planned.v1) {
-      segment->planned.accel = 0;
-      segment->planned.v0 = segment->planned.v1;
-      speed = segment->planned.v2;
-      continue;
-    }
-
-    // We have acceleration.
-    // Let's update the last invariant segment.
-    last_planned_invariant_segment_ = buffer_index > 0 ? buffer_index : 0;
-
-    const uint32_t steps = segment->planned.TotalSteps();
-    const uint32_t steps_to_vmax = uniformly_accelerated_ramp_distance(
-      segment_start_speed, segment->planned.v1, segment->target.accel);
-
-    // We intersect with the travel part.
-    if (steps_to_vmax <= segment->planned.travel) {
-      segment->planned.v0 = segment_start_speed;
-      segment->planned.accel = steps_to_vmax;
-      segment->planned.travel = segment->planned.travel - steps_to_vmax;
-      speed = segment->planned.v2;
-      continue;
+    next_segment = planning_buffer_[buffer_index];
+    if (ForwardUpdateSegment(previous_segment, next_segment) && !first_accel) {
+      // We have acceleration.
+      // Let's update the last invariant segment.
+      last_planned_invariant_segment_ = buffer_index;
```

Comment:
The name `last_planned_invariant_segment` is not really conveying much information for the reader. We should brainstorm for a better name. It should highlight that this is the cut-off point for a fully-planned output.

(also the question: when it is fully planned, could we use that opportunity to flush the buffer to the underlying hardware at this point ?)

Reply:
(solved): using num_segments_ready.